clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -338,6 +383,175 @@ impl FunctionType {
+            inner_type_check_type(
+                db,
+                contract_context,
+                &atom_inner_types.0,
+                &expected_inner_types.0,
+            )?;
+            inner_type_check_type(
+                db,
+                contract_context,
+                &atom_inner_types.1,
+                &expected_inner_types.1,
+            )?;
+        }
+        (
+            TypeSignature::SequenceType(SequenceSubtype::ListType(atom_list_type)),
+            TypeSignature::SequenceType(SequenceSubtype::ListType(expected_list_type)),
+        ) => {
+            inner_type_check_type(
+                db,
+                contract_context,
+                atom_list_type.get_list_item_type(),
+                expected_list_type.get_list_item_type(),
+            )?;
+        }
+        (
+            TypeSignature::TupleType(atom_tuple_type),
+            TypeSignature::TupleType(expected_tuple_type),
+        ) => {
+            for (name, atom_field_type) in atom_tuple_type.get_type_map() {
+                match expected_tuple_type.field_type(name) {
+                    Some(expected_field_type) => {
+                        inner_type_check_type(
+                            db,
+                            contract_context,
+                            atom_field_type,
+                            expected_field_type,
+                        )?;
+                    }
+                    None => {
+                        return Err(CheckErrors::TypeError(
+                            expected_type.clone(),
+                            atom_field_type.clone(),
+                        )
+                        .into())
+                    }
+                }
+            }
+        }
+        (
+            TypeSignature::TraitReferenceType(atom_trait_id),
+            TypeSignature::TraitReferenceType(expected_trait_id),
+        ) => {
+            if atom_trait_id != expected_trait_id {
+                let atom_trait = lookup_trait(db, contract_context, &atom_trait_id)?;
+                let expected_trait = lookup_trait(db, contract_context, expected_trait_id)?;
+                trait_check_trait_compliance(
+                    db,
+                    contract_context,
+                    &atom_trait_id,
+                    &atom_trait,
+                    expected_trait_id,
+                    &expected_trait,
+                )?;
+            }
+        }
+        (TypeSignature::NoType, _) => (),
+        (_, _) => {
+            if !expected_type.admits_type(&actual_type) {
+                return Err(
+                    CheckErrors::TypeError(expected_type.clone(), actual_type.clone()).into(),
+                );
+            }
+        }
+    }
+    Ok(expected_type.clone())
+}
+
+fn lookup_trait(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    trait_id: &TraitIdentifier,
+) -> CheckResult<BTreeMap<ClarityName, FunctionSignature>> {
```

Comment:
Should there be an analysis cost assessed here?  Asking because I don't see it happen at the call site.